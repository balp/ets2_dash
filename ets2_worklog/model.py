import datetime
import json
from dataclasses import dataclass
from typing import Optional, Dict, List

import ets2.model


@dataclass
class Delivered:
    auto_load: bool
    auto_park: bool
    cargo_damage: float
    time: int
    distance: float
    xp: int
    revenue: int


def delivered_from_dict(data: Dict) -> Delivered:
    return Delivered(auto_load=data['auto.load.used'], auto_park=data['auto.park.used'],
                     cargo_damage=data['cargo.damage'], time=data['delivery.time'],
                     distance=data['distance.km'], xp=data['earned.xp'], revenue=data['revenue'])


@dataclass
class Cancelled:
    penalty: int


def cancelled_from_dict(data: Dict) -> Cancelled:
    return Cancelled(penalty=data['cancel.penalty'])


@dataclass
class Job:
    config: Optional[ets2.model.JobConfig]
    started: Optional[datetime.datetime]
    ended: Optional[datetime.datetime]
    delivered: Optional[Delivered]
    cancelled: Optional[Cancelled]
    track: ets2.model.Tracks


class WorkLog:
    """A log work jobs in ETS2/ATS"""

    def __init__(self, data: ets2.model.Model):
        self._model = data
        self._model.register_observer(self)
        self._jobs: List[Job] = []

    def __repr__(self):
        return str(self._jobs)

    def notify(self, model: ets2.model.Model, change: str):
        if self._jobs:
            if model.job != self._jobs[-1].config:
                self._jobs[-1].ended = self.time_in_game(model)
                job = self.job_from_model(model)
                self._jobs.append(job)
            else:
                if model.telematic:
                    self._jobs[-1].track.add_telematic(model.telematic)
        else:
            job = self.job_from_model(model)
            self._jobs.append(job)

    def job_from_model(self, model):
        started = self.time_in_game(model)
        job = Job(config=model.job,
                  started=started,
                  ended=None,
                  delivered=None,
                  cancelled=None,
                  track=ets2.model.Tracks())
        return job

    def time_in_game(self, model):
        if model.telematic:
            started = datetime.datetime.fromordinal(1) \
                      + datetime.timedelta(minutes=model.telematic.common.game_time)
        else:
            started = None
        return started

    def job_delivered(self, delivered: Delivered) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].delivered = delivered

    def job_cancelled(self, cancelled: Cancelled) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].cancelled = cancelled


def add_json_to_work_log(work_log: WorkLog, json_data: json, topic: str) -> None:
    if topic == "ets2/info/gameplay/job.cancelled":
        work_log.job_cancelled(cancelled_from_dict(json_data))
    elif topic == "ets2/info/gameplay/job.delivered":
        work_log.job_delivered(delivered_from_dict(json_data))
