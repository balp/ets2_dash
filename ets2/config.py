#  Copyright (c) 2020. Anders Arnholm <Anders@Arnholm.se>
#
#   Permission to use, copy, modify, and distribute this software for any
#   purpose with or without fee is hereby granted, provided that the above
#   copyright notice and this permission notice appear in all copies.
#
#   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
#   WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
#   MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
#   ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
#   WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
#   ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
#   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
from dataclasses import dataclass
from typing import Dict, Optional

from ets2.types import Vector, vector_from_dict


@dataclass
class TruckConfig:
    id: str
    name: str
    adblue_capacity: float
    adblue_warning_factor: float
    battery_voltage_warning: float
    brake_air_pressure_emergency: float
    brake_air_pressure_warning: float
    brand: str
    brand_id: str
    cabin_position: Vector
    differential_ratio: float
    forward_ratio: float
    fuel_capacity: float
    fuel_warning_factor: float
    gears_forward: int
    gears_reverse: int
    head_position: Vector
    hook_position: Vector
    license_plate: str
    license_plate_country: str
    license_plate_country_id: str
    oil_pressure_warning: float
    retarder_steps: int
    reverse_ratio: float
    rpm_limit: float
    water_temperature_warning: float
    wheel_position: Vector
    wheel_powered: bool
    wheel_radius: float
    wheel_simulated: bool
    wheel_steerable: bool
    wheels_count: bool


def truckconfig_from_dict(data: Dict) -> Optional[TruckConfig]:
    if len(data):
        return TruckConfig(
            id=data['id'],
            name=data['name'],
            adblue_capacity=data['adblue.capacity'],
            adblue_warning_factor=data['adblue.warning.factor'],
            battery_voltage_warning=data['battery.voltage.warning'],
            brake_air_pressure_emergency=data['brake.air.pressure.emergency'],
            brake_air_pressure_warning=data['brake.air.pressure.warning'],
            brand=data['brand'],
            brand_id=data['brand_id'],
            cabin_position=vector_from_dict(data['cabin.position']),
            differential_ratio=data['differential.ratio'],
            forward_ratio=data['forward.ratio'],
            fuel_capacity=data['fuel.capacity'],
            fuel_warning_factor=data['fuel.warning.factor'],
            gears_forward=data['gears.forward'],
            gears_reverse=data['gears.reverse'],
            head_position=vector_from_dict(data['head.position']),
            hook_position=vector_from_dict(data['hook.position']),
            license_plate=data['license.plate'],
            license_plate_country=data['license.plate.country'],
            license_plate_country_id=data['license.plate.country.id'],
            oil_pressure_warning=data['oil.pressure.warning'],
            retarder_steps=data['retarder.steps'],
            reverse_ratio=data['reverse.ratio'],
            rpm_limit=data['rpm.limit'],
            water_temperature_warning=data['water.temperature.warning'],
            wheel_position=vector_from_dict(data['wheel.position']),
            wheel_powered=data['wheel.powered'],
            wheel_radius=data['wheel.radius'],
            wheel_simulated=data['wheel.simulated'],
            wheel_steerable=data['wheel.steerable'],
            wheels_count=data['wheels.count'],
        )
    else:
        return None


@dataclass
class TrailerConfig:
    id: str
    body_type: Optional[str]
    chain_type: Optional[str]
    cargo_accessory_id: str
    hook_position: Vector
    license_plate: str
    license_plate_country: str
    license_plate_country_id: str
    wheel_position: Vector
    wheel_powered: bool
    wheel_radius: float
    wheel_simulated: bool
    wheel_steerable: bool
    wheels_count: int


def trailer_config_from_dict(data: Dict) -> Optional[TrailerConfig]:
    if len(data):
        return TrailerConfig(
            id=data['id'],
            body_type=(data['body.type'] if 'body.type' in data else None),
            chain_type=(data['chain.type'] if 'chain.type' in data else None),
            cargo_accessory_id=data['cargo.accessory.id'],
            hook_position=vector_from_dict(data['hook.position']),
            license_plate=data['license.plate'],
            license_plate_country=data['license.plate.country'],
            license_plate_country_id=data['license.plate.country.id'],
            wheel_position=vector_from_dict(data['wheel.position']),
            wheel_powered=data['wheel.powered'],
            wheel_radius=data['wheel.radius'],
            wheel_simulated=data['wheel.simulated'],
            wheel_steerable=data['wheel.steerable'],
            wheels_count=data['wheels.count'],
        )
    else:
        return None