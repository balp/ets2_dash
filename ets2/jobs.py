from dataclasses import dataclass
from typing import Optional, Dict

import ets2.model


@dataclass
class Delivered:
    auto_load: bool
    auto_park: bool
    cargo_damage: float
    time: int
    distance: float
    xp: int
    revenue: int


def delivered_from_dict(data: Dict) -> Delivered:
    return Delivered(auto_load=data['auto.load.used'], auto_park=data['auto.park.used'],
                     cargo_damage=data['cargo.damage'], time=data['delivery.time'],
                     distance=data['distance.km'], xp=data['earned.xp'], revenue=data['revenue'])


@dataclass
class Cancelled:
    penalty: int


def cancelled_from_dict(data: Dict) -> Cancelled:
    return Cancelled(penalty=data['cancel.penalty'])


@dataclass
class Job:
    config: Optional[ets2.model.JobConfig]
    started: Optional[int]
    ended: Optional[int]
    delivered: Optional[Delivered]
    cancelled: Optional[Cancelled]
    track: ets2.model.Tracks
    id: Optional[int]
