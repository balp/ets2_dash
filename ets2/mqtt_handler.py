import json
import threading
import typing
from dataclasses import dataclass

from paho.mqtt import client as mqtt

from ets2.model import Model, add_json_to_model
from ets2.work_log import WorkLog, add_json_to_work_log


@dataclass
class GlobalState:
    active: bool = True


def on_message(_: typing.Any, userdata: typing.Any, message: typing.Any):
    """Handle a mqtt message and send to Model"""
    model, work_log = userdata
    assert isinstance(model, Model)
    json_data = json.loads(message.payload)
    add_json_to_model(model=model, json_data=json_data, topic=message.topic)
    if work_log is not None:
        assert isinstance(work_log, WorkLog)
        add_json_to_work_log(work_log=work_log, json_data=json_data,
                             topic=message.topic)


def mqtt_thread_loop(model: Model, work_log: WorkLog, state: GlobalState) -> None:
    client = mqtt.Client("ets2_gui", userdata=(model, work_log))
    client.on_message = on_message
    client.connect("localhost")
    client.subscribe("ets2/data")
    client.subscribe("ets2/game")
    client.subscribe("ets2/info")
    client.subscribe("ets2/info/config/job")
    client.subscribe("ets2/info/config/substances")
    # client.subscribe("ets2/info/config/controls")
    # client.subscribe("ets2/info/config/hshifter")
    client.subscribe("ets2/info/config/truck")
    # client.subscribe("ets2/info/config/trailer")

    client.subscribe("ets2/info/config/trailer.0")
    client.subscribe("ets2/info/config/trailer.1")
    client.subscribe("ets2/info/config/trailer.2")
    client.subscribe("ets2/info/config/trailer.3")
    client.subscribe("ets2/info/config/trailer.4")
    client.subscribe("ets2/info/config/trailer.5")
    client.subscribe("ets2/info/config/trailer.6")
    client.subscribe("ets2/info/config/trailer.7")
    client.subscribe("ets2/info/config/trailer.8")
    client.subscribe("ets2/info/config/trailer.9")

    client.subscribe("ets2/info/gameplay/job.cancelled")
    client.subscribe("ets2/info/gameplay/job.delivered")
    client.subscribe("ets2/info/gameplay/player.tollgate.paid")
    client.subscribe("ets2/info/gameplay/player.use.ferry")

    while state.active:
        client.loop(timeout=1.0)


def mqtt_model_handler():
    model: Model = Model()
    work_log: WorkLog = WorkLog(model, database=None)
    state: GlobalState = GlobalState()
    mqtt_reader_thread = threading.Thread(target=mqtt_thread_loop,
                                          args=(model, work_log, state))
    mqtt_reader_thread.start()
    return model, mqtt_reader_thread, state, work_log


def mqtt_event_loop(hmi, mqtt_reader_thread, state):
    while True:
        event, values = hmi.window.Read(timeout=50)
        hmi.update_data()
        if event == 'Exit':
            break
    state.active = False
    mqtt_reader_thread.join(timeout=2.0)
    hmi.window.CloseNonBlockingForm()