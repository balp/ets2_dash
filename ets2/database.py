import sqlite3

from ets2.jobs import Job


class DataBase:
    def __init__(self, db_path, db_name):
        db_path.mkdir(parents=True, exist_ok=True)
        database = db_path / db_name
        print(f"DataBase.__init__(... {database.absolute()} ...)")
        self._conn = sqlite3.connect(database.absolute(), isolation_level=None)
        cursor = self._conn.cursor()
        cursor.executescript("""
            CREATE TABLE IF NOT EXISTS job
            (
                id      integer primary key,
                started integer,
                ended   integer
            );
            
            CREATE TABLE IF NOT EXISTS job_config
            (
                id                     integer primary key,
                cargo                  text,
                cargo_id               text,
                cargo_mass             real,
                delivery_time          integer,
                destination_city       text,
                destination_city_id    text,
                destination_company    text,
                destination_company_id text,
                income                 integer,
                source_city            text,
                source_city_id         text,
                source_company         text,
                source_company_id      text,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
            
            CREATE TABLE IF NOT EXISTS job_delivered
            (
                id           integer primary key,
                auto_load    integer,
                auto_park    integer,
                cargo_damage real,
                time         integer,
                distance     real,
                xp           integer,
                revenue      integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
            
            CREATE TABLE IF NOT EXISTS job_cancelled
            (
                id      integer primary key,
                penalty integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
            
            CREATE TABLE IF NOT EXISTS tracks
            (
                id    integer primary key,
                last_time integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
            
            CREATE TABLE IF NOT EXISTS track_point
            (
                id        integer,
                count     integer,
                x real,
                y real,
                z real,
                PRIMARY KEY (id, count),
                FOREIGN KEY (id)
                    REFERENCES tracks (id)
            );
""")
        cursor.close()

    def save_job(self, job: Job):
        cursor = self._conn.cursor()
        if job.id is not None:
            cursor.execute("update job set started = ?, ended = ? where id = ?",
                           (job.started, job.ended, job.id))
        else:
            cursor.execute("insert into job (started, ended) values (?,?)",
                           (job.started, job.ended))
            job.id = cursor.lastrowid
        if job.config is not None:
            cursor.execute("replace into job_config (id, cargo, cargo_id, cargo_mass, delivery_time, "
                           "                         destination_city, destination_city_id, destination_company,"
                           "                         destination_company_id, income, source_city, source_city_id,"
                           "                         source_company, source_company_id)"
                           "   values (?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
                           (job.id,
                            job.config.cargo,
                            job.config.cargo_id,
                            job.config.cargo_mass,
                            job.config.delivery_time,
                            job.config.destination_city,
                            job.config.destination_city_id,
                            job.config.destination_company,
                            job.config.destination_company_id,
                            job.config.income,
                            job.config.source_city,
                            job.config.source_city_id,
                            job.config.source_company,
                            job.config.source_company_id))
        if job.delivered is not None:
            cursor.execute("replace into job_delivered ("
                           "    id, auto_load, auto_park, cargo_damage,"
                           "    time, distance, xp, revenue)"
                           "   values (?,?,?,?,?,?,?,?)",
                           (job.id,
                            job.delivered.auto_load,
                            job.delivered.auto_park,
                            job.delivered.cargo_damage,
                            job.delivered.time,
                            job.delivered.distance,
                            job.delivered.xp,
                            job.delivered.revenue))
        if job.cancelled is not None:
            cursor.execute("replace into job_cancelled (id, penalty) values (?,?)",
                           (job.id, job.cancelled.penalty))
        cursor.execute("replace into tracks (id, last_time) values (?,?)", (job.id, job.track.last_time))
        for i, point in enumerate(job.track.points):
            cursor.execute("replace into track_point (id, count, x, y, z) values (?,?,?,?,?)",
                           (job.id, i, point.position.x, point.position.y, point.position.z))
        cursor.close()
