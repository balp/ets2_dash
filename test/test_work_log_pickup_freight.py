import bz2
import json

import jsonpickle as jsonpickle
import pytest
from approvaltests import verify_with_namer, get_default_reporter
from approvaltests.core.namer import Namer

from ets2.model import Model, add_json_to_model
from ets2_worklog.model import WorkLog

def test_pickup_freight():
    model = Model()
    work_log: WorkLog = WorkLog(model)

    with bz2.open("data/ats_start_port_ang_coos_bay.mqtt.json.bz2") as gj:
        for line in gj:
            topic, json_str = line.decode('utf8').split(" ", 1)
            json_data = json.loads(json_str)
            add_json_to_model(model, json_data, topic)
    jsonpickle.set_encoder_options('json', sort_keys=True, indent=4)
    jsonpickle.set_preferred_backend('json')
    verify_with_namer(data=jsonpickle.encode(work_log),
                      namer=Namer(extension='.json'),
                      reporter=get_default_reporter())

def test_deliver_freight():
    model = Model()
    work_log: WorkLog = WorkLog(model)
    for f in ["data/ats_start_port_ang_coos_bay.mqtt.json.bz2",
              "data/ats_end_port_ang_coos_bay.mqtt.json.bz2"]:
        with bz2.open(f) as gj:
            for line in gj:
                topic, json_str = line.decode('utf8').split(" ", 1)
                json_data = json.loads(json_str)
                add_json_to_model(model, json_data, topic)
        jsonpickle.set_encoder_options('json', sort_keys=True, indent=4)
    jsonpickle.set_preferred_backend('json')
    verify_with_namer(data=jsonpickle.encode(work_log),
                      namer=Namer(extension='.json'),
                      reporter=get_default_reporter())



