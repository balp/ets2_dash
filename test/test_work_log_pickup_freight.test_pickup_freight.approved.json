{
    "_db": {
        "_conn": {
            "py/object": "sqlite3.Connection"
        },
        "py/object": "ets2.database.DataBase"
    },
    "_model": {
        "_observers": [
            {
                "py/id": 0
            }
        ],
        "game": {
            "id": "ats",
            "name": "American Truck Simulator 1.36.1.34s",
            "py/object": "ets2.model.Game",
            "raw_version": 65539,
            "version": {
                "major": 1,
                "minor": 3,
                "py/object": "ets2.model.Version"
            }
        },
        "info": {
            "paused": false,
            "py/object": "ets2.model.Info"
        },
        "job": {
            "cargo": "Lumber",
            "cargo_id": "lumber_b",
            "cargo_mass": 21387.240234375,
            "delivery_time": 38821,
            "destination_city": "Coos Bay",
            "destination_city_id": "coos_bay",
            "destination_company": "Olthon Homes",
            "destination_company_id": "oh_con_hom",
            "income": 18742,
            "py/object": "ets2.model.JobConfig",
            "source_city": "Port Angeles",
            "source_city_id": "port_angeles",
            "source_company": "Deepgrove",
            "source_company_id": "dg_wd_saw"
        },
        "py/object": "ets2.model.Model",
        "telematic": {
            "common": {
                "game_time": 37228,
                "py/object": "ets2.model.Common",
                "rest_stop": 220,
                "scale": 3.0
            },
            "py/object": "ets2.model.Telematic",
            "trailer": {
                "acceleration_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": -0.007289553992450237,
                    "y": 3.3804564736783504e-05,
                    "z": -0.001828339183703065
                },
                "acceleration_linear": {
                    "py/object": "ets2.model.Vector",
                    "x": -0.007743027992546558,
                    "y": -0.024335527792572975,
                    "z": -0.006259627174586058
                },
                "cargo_damage": 0.0,
                "connected": false,
                "py/object": "ets2.model.Trailer",
                "velocity_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": -0.0010362350149080157,
                    "y": 1.0830994142452255e-05,
                    "z": -0.0002931451308541
                },
                "velocity_linear": {
                    "py/object": "ets2.model.Vector",
                    "x": -0.0008678262238390744,
                    "y": -0.006193636916577816,
                    "z": -0.0012792126508429646
                },
                "wear_chassis": 0.0,
                "wear_wheels": 0.0,
                "world_placement": {
                    "orientation": {
                        "heading": 0.9233342409133911,
                        "pitch": -0.0002509714977350086,
                        "py/object": "ets2.model.Euler",
                        "roll": -1.7872327589429915e-05
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": -104230.42456054688,
                        "y": 12.102143287658691,
                        "z": -66722.82945251465
                    },
                    "py/object": "ets2.model.Placement"
                }
            },
            "truck": {
                "adblue": 79.88677978515625,
                "adblue_warning": false,
                "battery_voltage": 13.751858711242676,
                "battery_voltage_warning": false,
                "brake_air_pressure": 101.76799774169922,
                "brake_air_pressure_emergency": false,
                "brake_air_pressure_warning": false,
                "brake_motor": false,
                "brake_parking": false,
                "brake_retarder": 0,
                "brake_temperature": 133.95758056640625,
                "cabin_acceleration_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0,
                    "y": 0.0,
                    "z": 0.0
                },
                "cabin_offset": {
                    "orientation": {
                        "heading": 0.0,
                        "pitch": 0.0,
                        "py/object": "ets2.model.Euler",
                        "roll": 0.0
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": 0.0,
                        "y": 0.0,
                        "z": 0.0
                    },
                    "py/object": "ets2.model.Placement"
                },
                "cabin_velocity_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0,
                    "y": 0.0,
                    "z": 0.0
                },
                "cruise_control": 0.0,
                "dashboard_backlight": 1.0,
                "displayed_gear": 3,
                "effective_brake": 1.0,
                "effective_clutch": 0.22759786248207092,
                "effective_steering": -0.03430912271142006,
                "effective_throttle": 0.0,
                "electric_enabled": true,
                "engine_enabled": true,
                "engine_gear": 0,
                "engine_rpm": 600.1014404296875,
                "fuel_amount": 755.7356567382812,
                "fuel_consumption_average": 0.459848552942276,
                "fuel_range": 2373.0830078125,
                "fuel_warning": false,
                "head_offset": {
                    "orientation": {
                        "heading": 0.0,
                        "pitch": -0.01944444701075554,
                        "py/object": "ets2.model.Euler",
                        "roll": 0.0
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": 0.0,
                        "y": -0.019999999552965164,
                        "z": 0.0
                    },
                    "py/object": "ets2.model.Placement"
                },
                "hshifter_slot": 0,
                "input_brake": 0.0,
                "input_clutch": 0.0,
                "input_steering": -0.03430912271142006,
                "input_throttle": 0.0,
                "lblinker": false,
                "light_aux_front": 0,
                "light_aux_roof": 0,
                "light_beacon": false,
                "light_beam_high": false,
                "light_beam_low": false,
                "light_brake": false,
                "light_lblinker": false,
                "light_parking": false,
                "light_rblinker": false,
                "light_reverse": false,
                "local_acceleration_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": -0.009236913174390793,
                    "y": 5.896131551708095e-05,
                    "z": -0.00035891516017727554
                },
                "local_acceleration_linear": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.00040731264743953943,
                    "y": -0.10982277244329453,
                    "z": 0.013066155835986137
                },
                "local_velocity_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.00017813759041018784,
                    "y": -2.316460268048104e-05,
                    "z": -2.6898604119196534e-05
                },
                "local_velocity_linear": {
                    "py/object": "ets2.model.Vector",
                    "x": -0.00013748378842137754,
                    "y": -0.006711939349770546,
                    "z": -0.0002523705188650638
                },
                "navigation_distance": 0.0,
                "navigation_speed_limit": 15.646400451660156,
                "navigation_time": 0.0,
                "odometer": 16461.01171875,
                "oil_pressure": 42.539058685302734,
                "oil_pressure_warning": false,
                "oil_temperature": 79.49394226074219,
                "py/object": "ets2.model.Truck",
                "rblinker": false,
                "speed": 0.00036216891021467745,
                "water_temperature": 59.09098434448242,
                "water_temperature_warning": false,
                "wear_cabin": 0.0,
                "wear_chassis": 0.0,
                "wear_engine": 0.0007700005080550909,
                "wear_transmission": 0.0007700005080550909,
                "wear_wheels": 0.007700022310018539,
                "wipers": false,
                "world_placement": {
                    "orientation": {
                        "heading": 0.9165711998939514,
                        "pitch": 1.4115723388385959e-05,
                        "py/object": "ets2.model.Euler",
                        "roll": 1.13836483706109e-06
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": -104272.02157592773,
                        "y": 12.105117797851562,
                        "z": -66604.86675262451
                    },
                    "py/object": "ets2.model.Placement"
                }
            }
        },
        "tracks": {
            "last_time": 37228,
            "points": [
                {
                    "orientation": {
                        "heading": 0.4340660274028778,
                        "pitch": 0.0,
                        "py/object": "ets2.model.Euler",
                        "roll": 0.0
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": -104305.18292236328,
                        "y": 12.102846145629883,
                        "z": -66655.20185089111
                    },
                    "py/object": "ets2.model.Placement"
                },
                {
                    "orientation": {
                        "heading": 0.9165719151496887,
                        "pitch": -4.098113549844129e-06,
                        "py/object": "ets2.model.Euler",
                        "roll": 2.504402573322295e-06
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": -104272.02157592773,
                        "y": 12.105484962463379,
                        "z": -66604.86669540405
                    },
                    "py/object": "ets2.model.Placement"
                }
            ],
            "py/object": "ets2.model.Tracks"
        },
        "trailer_config": [
            {
                "body_type": "flatbed",
                "cargo_accessory_id": "scs_flatbed.beams_28",
                "chain_type": "double",
                "hook_position": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0,
                    "y": 1.0,
                    "z": -5.362109661102295
                },
                "id": "scs_flatbed.flatbed28_hookx2esii",
                "license_plate": "40118TR",
                "license_plate_country": "Washington",
                "license_plate_country_id": "washington",
                "py/object": "ets2.model.TrailerConfig",
                "wheel_position": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.9244899749755859,
                    "y": 0.5199999809265137,
                    "z": 1.309550404548645
                },
                "wheel_powered": false,
                "wheel_radius": 0.5199999809265137,
                "wheel_simulated": true,
                "wheel_steerable": false,
                "wheels_count": 2
            },
            {
                "body_type": null,
                "cargo_accessory_id": "",
                "chain_type": null,
                "hook_position": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0,
                    "y": 1.0,
                    "z": -1.660975694656372
                },
                "id": "scs_flatbed.dolly_cx2esii",
                "license_plate": "10960TR",
                "license_plate_country": "Washington",
                "license_plate_country_id": "washington",
                "py/object": "ets2.model.TrailerConfig",
                "wheel_position": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.9244899749755859,
                    "y": 0.5199999809265137,
                    "z": 0.9275853037834167
                },
                "wheel_powered": false,
                "wheel_radius": 0.5199999809265137,
                "wheel_simulated": true,
                "wheel_steerable": false,
                "wheels_count": 2
            },
            {
                "body_type": null,
                "cargo_accessory_id": "scs_flatbed.beams_28",
                "chain_type": null,
                "hook_position": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0,
                    "y": 1.0,
                    "z": -5.362109661102295
                },
                "id": "scs_flatbed.flatbed28_hookx2esii",
                "license_plate": "20071TR",
                "license_plate_country": "Washington",
                "license_plate_country_id": "washington",
                "py/object": "ets2.model.TrailerConfig",
                "wheel_position": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.9244899749755859,
                    "y": 0.5199999809265137,
                    "z": 1.309550404548645
                },
                "wheel_powered": false,
                "wheel_radius": 0.5199999809265137,
                "wheel_simulated": true,
                "wheel_steerable": false,
                "wheels_count": 2
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null
        ],
        "truck_config": {
            "adblue_capacity": 80.0,
            "adblue_warning_factor": 0.15000000596046448,
            "battery_voltage_warning": 11.880000114440918,
            "brake_air_pressure_emergency": 29.579999923706055,
            "brake_air_pressure_warning": 59.15999984741211,
            "brand": "Volvo",
            "brand_id": "volvo",
            "cabin_position": {
                "py/object": "ets2.model.Vector",
                "x": 0.0,
                "y": 3.0,
                "z": -2.0
            },
            "differential_ratio": 3.549999952316284,
            "forward_ratio": 0.7300000190734863,
            "fuel_capacity": 758.0,
            "fuel_warning_factor": 0.15000000596046448,
            "gears_forward": 13,
            "gears_reverse": 3,
            "head_position": {
                "py/object": "ets2.model.Vector",
                "x": -0.5642366409301758,
                "y": -0.5814938545227051,
                "z": 0.570602536201477
            },
            "hook_position": {
                "py/object": "ets2.model.Vector",
                "x": 0.0,
                "y": 1.0,
                "z": 1.068235158920288
            },
            "id": "vehicle.volvo.vnl",
            "license_plate": "8A7P001",
            "license_plate_country": "California",
            "license_plate_country_id": "california",
            "name": "VNL",
            "oil_pressure_warning": 10.149999618530273,
            "py/object": "ets2.model.TruckConfig",
            "retarder_steps": 3,
            "reverse_ratio": -4.03000020980835,
            "rpm_limit": 2100.0,
            "water_temperature_warning": 105.0,
            "wheel_position": {
                "py/object": "ets2.model.Vector",
                "x": 0.9612910747528076,
                "y": 0.5059999823570251,
                "z": 0.5031635761260986
            },
            "wheel_powered": true,
            "wheel_radius": 0.5072842240333557,
            "wheel_simulated": true,
            "wheel_steerable": false,
            "wheels_count": 6
        }
    },
    "jobs": [
        {
            "cancelled": null,
            "config": null,
            "delivered": null,
            "ended": 37228,
            "id": 1,
            "py/object": "ets2.jobs.Job",
            "started": null,
            "track": {
                "last_time": 37228,
                "points": [
                    {
                        "py/id": 37
                    },
                    {
                        "py/id": 40
                    }
                ],
                "py/object": "ets2.model.Tracks"
            }
        },
        {
            "cancelled": null,
            "config": {
                "py/id": 8
            },
            "delivered": null,
            "ended": null,
            "id": 2,
            "py/object": "ets2.jobs.Job",
            "started": 37228,
            "track": {
                "last_time": 37228,
                "points": [
                    {
                        "py/id": 40
                    }
                ],
                "py/object": "ets2.model.Tracks"
            }
        }
    ],
    "py/object": "ets2.work_log.WorkLog"
}